### PR TITLE
Adding "susceptibility" or "social-influence"

### PR DESCRIPTION
**Pull request by jig08**
... wrt **Issue #85 Agent Social Influence**

* Setting `use_susceptibility: true` in `DEFAULT.yaml`

* Running commands `./build.sh` and then `./run.sh`

Sample file in the `output` folder i.e. `change_in_ideology_step_0.dat` showing changes in agent ideologies in step_0.

```

n_steps: 0

agent_id: 0
agent_ideology_old: 3
agent_ideology_new: 3

agent_id: 1
agent_ideology_old: 1
agent_ideology_new: 3

agent_id: 2
agent_ideology_old: 3
agent_ideology_new: 3

agent_id: 3
agent_ideology_old: 2
agent_ideology_new: 3

agent_id: 4
agent_ideology_old: 2
agent_ideology_new: 3

agent_id: 5
agent_ideology_old: 3
agent_ideology_new: 3

.....
```

### CODE REVIEW

File: src/analyzer_main.cpp [hunk 1]
```diff
@@ -30,6 +30,7 @@
 #include <iomanip>
 #include <cmath>
 #include <deque>
+#include <cassert>
```
Comment (ludamad) [inline src/analyzer_main.cpp]: We should not need <cassert>

File: src/analyzer_main.cpp [hunk 2]
```diff
@@ -559,6 +560,58 @@ struct Analyzer {
             return false;
         }
 
+        //Handling susceptibility
+        if(config.use_susceptibility == true && stats.n_steps % 10 == 0){
```
Comment (ludamad) [inline src/analyzer_main.cpp]: == true is superfluous

File: src/analyzer_main.cpp [hunk 3]
```diff
@@ -559,6 +560,58 @@ struct Analyzer {
...
+            myfile << "\nn_steps: " << stats.n_steps;
+
+            int old_ideology;
+
+            for (Agent& agent : network) {
+                if (agent.susceptibility == 1.0) {
+
+                    myfile << "\n\nagent_id: " << agent.id;
+
+                    old_ideology = agent.ideology_bin;
```
Comment (ludamad) [inline src/analyzer_main.cpp]: 'int old_ideology' should be declared here

File: src/analyzer_main.cpp (hunk 2 above)
Comment (ludamad) [inline src/analyzer_main.cpp]: Try to copy the spacing in the rest of hashkat. `if (config...) {`

File: src/analyzer_main.cpp [hunk 4]
```diff
@@ -559,6 +560,58 @@ struct Analyzer {
             return false;
         }
 
+        //Handling susceptibility
+        if(config.use_susceptibility == true && stats.n_steps % 10 == 0){
+
+            ofstream myfile;
+
+            stringstream ss;
+            ss << "output/change_in_ideology_step_" << stats.n_steps << ".dat";
```
Comment (ludamad) [inline src/analyzer_main.cpp]: I'm not sure this output file is appropriate. I'd rather see one log of all ideology changes and the simulation time (not step) where they occur, and have that controlled by an output: variable

File: src/analyzer_main.cpp [hunk 5]
```diff
@@ -559,6 +560,58 @@ struct Analyzer {
...
+
+                    myfile << "\n\nagent_id: " << agent.id;
+
+                    old_ideology = agent.ideology_bin;                
+
+                    // Looking at the ideology of all of an agent's followings:
+                    int counts[N_BIN_IDEOLOGIES];
+                    for (int& count : counts) {
+                        count = 0; 
+                        // Counts start at 0
```
Comment (ludamad) [inline src/analyzer_main.cpp]: Nit: Comments before the code or on the same line. Eg `count = 0; // Counts start at 0`, but not on the next line.

File: src/analyzer_main.cpp [hunk 6]
```diff
@@ -559,6 +560,58 @@ struct Analyzer {
...
+                    }
+                    for (int following_id : agent.following_set.as_vector()) {
+                        Agent& following = network[following_id];
+                        counts[following.ideology_bin]++;
+                    }
+
+                    int most_common_ideology = -1;
+                    int max_count = -1;
+                    for(int i=0; i<N_BIN_IDEOLOGIES; i++){
```
Comment (ludamad) [inline src/analyzer_main.cpp]: Nit: Spacing again. You can get Eclipse to autoformat the spacing with ctrl+shift+f, there should be space like `i < N_BIN_IDEOLOGIES` and `for (int`.

File: src/config_dynamic.cpp [hunk 7]
```diff
@@ -357,6 +358,9 @@ static AgentTypeVector parse_agents_configuration(const Node& node) {
 		const Node& follow_rate = child["rates"]["follow"];
 		const Node& tweet_rate = child["rates"]["tweet"];
 
+        //Parsing Susceptibility
```
Comment (ludamad) [inline src/config_dynamic.cpp]: This comment is a bit lacking in oomph. The line it comments already has the words 'parse' and 'susceptibility'. I would remove it.

File: DEFAULT.yaml [hunk 8]
```diff
@@ -366,3 +370,5 @@ agents:
         follow: {function: constant, value: 0.000001}
         # Rate for tweets from this agent:
         tweet: {function: constant, value: 0.00001}
+    #Introducing susceptibility
```
Comment (ludamad) [inline DEFAULT.yaml]: This comment should describe what this represents

File: src/analyzer_main.cpp [hunk 9]
```diff
@@ -559,6 +560,58 @@ struct Analyzer {
             return false;
         }
 
+        //Handling susceptibility
+        if(config.use_susceptibility == true && stats.n_steps % 10 == 0){
+
+            ofstream myfile;
```
Comment (ludamad) [inline src/analyzer_main.cpp]: 'myfile' could be more descriptive

Comment (jig08) [inline src/analyzer_main.cpp]: Get it, shall change it.

File: DEFAULT.yaml (hunk 8 above)
Comment (jig08) [inline DEFAULT.yaml]: Okay @ludamad . Will add more details :)

File: src/config_dynamic.cpp (hunk 7 above)
Comment (jig08) [inline src/config_dynamic.cpp]: Removing this line altogether ?

File: src/analyzer_main.cpp (hunk 6 above)
Comment (jig08) [inline src/analyzer_main.cpp]: :+1:

File: src/analyzer_main.cpp (hunk 5 above)
Comment (jig08) [inline src/analyzer_main.cpp]: Gotcha.

File: src/analyzer_main.cpp (hunk 4 above)
Comment (jig08) [inline src/analyzer_main.cpp]: Umm, could you kindly elaborate ?

File: src/analyzer_main.cpp (hunk 2 above)
Comment (jig08) [inline src/analyzer_main.cpp]: >== true is superfluous

+1

File: src/analyzer_main.cpp (hunk 1 above)
Comment (jig08) [inline src/analyzer_main.cpp]: I'll check, it gave error afaik.

File: src/analyzer_main.cpp (hunk 4 above)
Comment (ludamad) [inline src/analyzer_main.cpp]: This isn't too critical, but the output will be very big in big networks. You can just print out the changed agents, or leave it for now.

File: src/config_dynamic.cpp (hunk 7 above)
Comment (ludamad) [inline src/config_dynamic.cpp]: Yes

File: src/analyzer_main.cpp (hunk 4 above)
Comment (jig08) [inline src/analyzer_main.cpp]: Leave any change, or leave printing altogether ?

File: src/config_dynamic.cpp (hunk 7 above)
Comment (jig08) [inline src/config_dynamic.cpp]: Ok.

File: src/analyzer_main.cpp (hunk 4 above)
Comment (ludamad) [inline src/analyzer_main.cpp]: I'm fine with anything you decide. Just noting my concern that it could be a lot of output.

Comment (jig08) [inline src/analyzer_main.cpp]: Agreed that it shall be trouble when we have a big network, though.

**Comment by jig08** [comment]
@ludamad Thanks, I'll push this into master then ?

**Comment by jig08** [comment]
@ludamad Thanks, Pushing it into master then.

**Pull request merged**